Common/src/geometry_structure.cpp

```diff
@@ -9386,1429 +9383,1580 @@ void CPhysicalGeometry::Read_CGNS_Format_Parallel(CConfig *config, string val_me
+                     MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+  SU2_MPI::Allreduce(&Local_nElemHex,     &Global_nelem_hexa,      1,
+                     MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+  SU2_MPI::Allreduce(&Local_nElemPrism,   &Global_nelem_prism,     1,
+                     MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+  SU2_MPI::Allreduce(&Local_nElemPyramid, &Global_nelem_pyramid,   1,
+                     MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+#else
+  Global_nelem_triangle = nelem_triangle;
+  Global_nelem_quad     = nelem_quad;
+  Global_nelem_tetra    = nelem_tetra;
+  Global_nelem_hexa     = nelem_hexa;
+  Global_nelem_prism    = nelem_prism;
+  Global_nelem_pyramid  = nelem_pyramid;
+#endif
+  
+  /*--- For now, the master node takes care of all markers. ---*/
+  
+  if (rank == MASTER_NODE) {
+    
+    /*--- Read number of markers ---*/
+    
+    nMarker = nMarkers;
+    cout << nMarker << " surface markers." << endl;
+    config->SetnMarker_All(nMarker);
+    bound = new CPrimalGrid**[nMarker];
+    nElem_Bound = new unsigned long [nMarker];
+    Tag_to_Marker = new string [nMarker_Max];
+    
+    iMarker = 0;
+    
+    for ( int s = 0; s < nsections; s++ ) {
+      if ( !isVolume[s] ) {
+        
+        /*--- Initialize some counter variables ---*/
+        
+        nelem_edge_bound = 0; nelem_triangle_bound = 0;
+        nelem_quad_bound = 0; iElem = 0;
+        
+        Marker_Tag = string(sectionNames[s].data());
+        
+        /*--- Remove whitespaces from the marker names ---*/
+        Marker_Tag.erase(remove(Marker_Tag.begin(), Marker_Tag.end(),' '),
+                         Marker_Tag.end());
+        
+        if (Marker_Tag != "SEND_RECEIVE") {
+          nElem_Bound[iMarker] = nElems[s];
+          if (rank == MASTER_NODE) {
+            cout << nElem_Bound[iMarker]  << " boundary elements in index ";
+            cout << iMarker <<" (Marker = " <<Marker_Tag<< ")." << endl;
           }
-
-          /*--- Free the memory for the conn. from the CGNS file. ---*/
-          
-          delete [] connElemCGNS;
-          delete [] isMixed;
-          
-          /*--- We now have the connectivity stored in linearly partitioned
-           chunks. We need to loop through and decide how many elements we
-           must send to each rank in order to have all elements that
-           surround a particular "owned" node on each rank (i.e., elements
-           will appear on multiple ranks). First, initialize a counter
-           and flag. ---*/
-          
-          int *nElem_Send = new int[size+1]; nElem_Send[0] = 0;
-          int *nElem_Recv = new int[size+1]; nElem_Recv[0] = 0;
-          int *nElem_Flag = new int[size];
-          
-          for (int ii=0; ii < size; ii++) {
-            nElem_Send[ii] = 0;
-            nElem_Recv[ii] = 0;
-            nElem_Flag[ii]= -1;
-          }
-          nElem_Send[size] = 0; nElem_Recv[size] = 0;
+          bound[iMarke
```

Comment:
We really need a 2D container class to standardize this type of stuff. Can you mark this as a ToDo or Fixme for when we do have such a container.

Reply:
Sorry @MicK7 I think git is showing a preview that does not match the code where I put the TODO comment, it is supposed to be line 10626 `passivedouble **gridCoords = new passivedouble*[val_ncoords];`

Reply:
Ok I will move it.

Do you think about a light container like this :

```
template<typename T>
struct arr2d {
size_t dim1;
size_t dim2;
T*  arr_;

arr2d(dim1, dim2){
  this->arr_ = new T[dim1*dim2];
  this->dim1 = dim1;
  this->dim2 = dim2;
}

~arr2d() { del [] this->arr_; }

inline T* value( idx1, idx2){ return &(this->arr_[idx1*dim2+idx2]);}

const T& value( idx1, idx2) const { return this->arr_[idx1*dim2+idx2];}

}
```

I think it could be useful in other area if someone wants to experiment SOA structures.

Reply:
Yep something of the sort, I am working on one as part of #716 .
For the exact purpose you mention, and maybe in the future support some basic matrix algebra operations that currently we have scattered throughout the code.

Reply:
On second thought if points coordinates were store in a contiguous array we would not even need this buffer as the CGNS library would do the mapping for us.

Reply:
Depending on whether we store by rows or columns :)

Reply:
This is not a problem. For instance, if points are stored in an interleaved fashion, cgns 3.4 can read directly to memory by using a stride argument (this is how it is done in ParaView/VTK to read CGNS data).